apps/prairielearn/src/lib/group-update.js

```diff
@@ -80,50 +77,30 @@ export async function uploadInstanceGroups(assessment_id, csvFile, user_id, auth
         const csvStream = streamifier.createReadStream(csvFile.buffer, {
           encoding: 'utf8',
         });
-        const csvConverter = csvtojson({
-          colParser: {
-            groupname: 'string',
-            groupName: 'string',
-            uid: 'string',
-            UID: 'string',
-            Uid: 'string',
-          },
-          maxRowLength: 10000,
-        });
-        const updateList = [];
-        await csvConverter.fromStream(csvStream).subscribe((json) => {
-          let groupName = json.groupName || json.groupname || null;
-          let uid = json.uid || json.UID || json.Uid || null;
-          updateList.push([groupName, uid]);
-        });
-        const result = await callRow(
-          'assessment_groups_update',
-          [assessment_id, updateList, authn_user_id],
-          AssessmentGroupsUpdateResultSchema,
-        );
-        let errorCount = 0;
-
-        const notExist = result.not_exist_user;
-        if (notExist) {
-          job.verbose(`----------------------------------------`);
-          job.verbose(`ERROR: The following users do not exist. Please check their uids first.`);
-          notExist.forEach((user) => {
-            job.verbose(user);
-          });
-          errorCount += notExist.length;
-        }
+        const csvConverter = csvtojson({ checkType: false, maxRowLength: 10000 });
+        let totalRows = 0,
+          successCount = 0;
+        await csvConverter.fromStream(csvStream).subscribe(async (row) => {
+          row = _.mapKeys(row, (_v, k) => k.toLowerCase());
+          const groupName = row.groupname || null;
+          const uid = row.uid || null;
+          // Ignore rows without a group name and uid (blank lines)
+          if (!uid && !groupName) return;
+
+          totalRows++;
+          const user = await selectUserByUid(uid);
+          if (!user) {
+            job.error(`User with uid ${uid} does not exist`);
+            return;
+          }
 
-        const inGroup = result.already_in_group;
-        if (inGroup) {
-          job.verbose(`----------------------------------------`);
-          job.verbose(`ERROR: The following users are already in a group.`);
-          inGroup.forEach((user) => {
-            job.verbose(user);
-          });
-          errorCount += inGroup.length;
-        }
+          await createOrAddToGroup(groupName, assessment_id, [user.user_id], authn_user_id).then(
+            () => successCount++,
+            (err) => job.error(`Error adding ${uid} to group ${groupName}: ${err.message}`),
+          );
```

Comment:
Given this is already inside a lock, is this really necessary? I'm particularly considering the fact that a failure in one row does not necessarily mean we would rollback the whole file. Note that the functions already use transactions, so for upload each row is a transaction, and for auto-groups each group is a transaction.

Reply:
The transaction would be for atomicity, which I think is a valuable property to have. While I agree that we wouldn't want an "expected" error on one row to roll back all the changes, the code isn't written such that we'd throw a transaction-aborting error in the normal course of operation. The only reason we would expect the transaction to fail is e.g. a server crash, in which case I think having this be all-or-nothing is the correct approach.

Reply:
Done in cea943e46b3975a1f5ed4dbeb01f76b7641224db (auto-groups) and 0b134356747eb27b184cdd60230ccafd6bd64518 (upload). I separated the UI part of the upload to retrieve the list of entries first, so that the UI section was not inside the transaction.